packages/world-modules/src/modules/erc20/ERC20System.sol

```diff
@@ -0,0 +1,288 @@
+  }
+
+  /**
+   * @dev Destroys a `value` amount of tokens from `account`, lowering the total supply.
+   * Relies on the `_update` mechanism.
+   *
+   * Emits a {Transfer} event with `to` set to the zero address.
+   *
+   * NOTE: This function is not virtual, {_update} should be overridden instead
+   */
+  function burn(address account, uint256 value) public {
+    // Require the caller to own the namespace
+    _requireOwner();
+
+    if (account == address(0)) {
+      revert ERC20InvalidSender(address(0));
+    }
+
+    _update(account, address(0), value);
+  }
+
+  /**
+   * @dev Moves a `value` amount of tokens from `from` to `to`.
+   *
+   * This internal function is equivalent to {transfer}, and can be used to
+   * e.g. implement automatic token fees, slashing mechanisms, etc.
+   *
+   * Emits a {Transfer} event.
+   *
+   * NOTE: This function is not virtual, {_update} should be overridden instead.
+   */
+  function _transfer(address from, address to, uint256 value) internal {
+    if (from == address(0)) {
+      revert ERC20InvalidSender(address(0));
+    }
+    if (to == address(0)) {
+      revert ERC20InvalidReceiver(address(0));
+    }
+    _update(from, to, value);
+  }
+
+  /**
+   * @dev Transfers a `value` amount of tokens from `from` to `to`, or alternatively mints (or burns) if `from`
+   * (or `to`) is the zero address. All customizations to transfers, mints, and burns should be done by overriding
+   * this function.
+   *
+   * Emits a {Transfer} event.
+   */
+  function _update(address from, address to, uint256 value) internal virtual {
+    bytes14 namespace = _namespace();
+    ResourceId metadataTableId = _metadataTableId(namespace);
+    ResourceId balanceTableId = _balancesTableId(namespace);
+
+    if (from == address(0)) {
+      // Overflow check required: The rest of the code assumes that totalSupply never overflows
+      Metadata.setTotalSupply(metadataTableId, Metadata.getTotalSupply(metadataTableId) + value);
+    } else {
+      uint256 fromBalance = Balances.get(balanceTableId, from);
+      if (fromBalance < value) {
+        revert ERC20InsufficientBalance(from, fromBalance, value);
+      }
+      unchecked {
+        // Overflow not possible: value <= fromBalance <= totalSupply.
+        Balances.set(balanceTableId, from, fromBalance - value);
+      }
+    }
+
+    if (to == address(0)) {
+      unchecked {
+        // Overflow not possible: value <= totalSupply or value <= fromBalance <= totalSupply.
+        Metadata.setTotalSupply(metadataTableId, Metadata.getTotalSupply(metadataTableId) - value);
+      }
+    } else {
+      unchecked {
+        // Overflow not possible: balance + value is at most totalSupply, which we know fits into a uint256.
+        Balances.set(balanceTableId, to, Balances.get(balanceTableId, to) + value);
+      }
+    }
+
+    // Emit Transfer event on puppet
+    puppet().log(IERC20Events.Transfer.selector, _toBytes32(from), _toBytes32(to), abi.encode(value));
```

Comment:
wtb solidity native `encodeLog` like `encodeCall`

Reply:
we should open an EIP for that, would be nice for sure